core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/authorization/PermissionTree.java

```diff
@@ -1054,6 +1062,14 @@ void updateResourcePermissionsById(String resourceId) throws UserStoreException
             } finally {
                 write.unlock();
             }
+
+            write.lock();
+            try {
+                this.root = tree.root;
```

Comment:
Would nt this better go to line 1062. Having separate write.lock() will cause race conditions.

Reply:
Yes that is true. Fixed it